discovery.go

```diff
@@ -212,31 +212,48 @@ func (r *Router) DiscoveryAllBuckets(ctx context.Context) error {
 	return nil
 }
 
-// startCronDiscovery is discovery_service_f analog with goroutines instead fibers
-func (r *Router) startCronDiscovery(ctx context.Context) error {
+// cronDiscovery is discovery_service_f analog with goroutines instead fibers
+func (r *Router) cronDiscovery(ctx context.Context) {
+	var iteration uint64
+
 	for {
 		select {
 		case <-ctx.Done():
 			r.metrics().CronDiscoveryEvent(false, 0, "ctx-cancel")
-
-			return ctx.Err()
+			r.log().Infof(ctx, "cron discovery has been stopped after %d iterations", iteration)
+			return
 		case <-time.After(r.cfg.DiscoveryTimeout):
-			r.log().Debugf(ctx, "started new cron discovery")
-
-			tStartDiscovery := time.Now()
+			iteration++
+		}
 
+		// Since the current for loop should not stop until ctx->Done() event fires,
+		// we should be able to continue execution even a panic occures.
+		// Therefore, we should wrap everyting into anonymous function that recovers after panic.
+		// (Similar to pcall in lua/tarantool)
+		func() {
 			defer func() {
-				r.log().Infof(ctx, "discovery done since %s", time.Since(tStartDiscovery))
+				if recovered := recover(); recovered != nil {
+					// Another one panic may happen due to log function below (e.g. bug in log().Errorf), in this case we have two options:
+					// 1. recover again and log nothing: panic will be muted and lost
+					// 2. don't try to recover, we hope that the second panic will be logged somehow by go runtime
+					// So, we choose the second behavior
+					r.log().Errorf(ctx, "ACHTUNG: something unexpected has happened in cronDiscovery(%d): %v", iteration, recovered)
```

Comment:
I think we shouldn't use the expression ACHTUNG and replace it with PANIC

It’s also worth logging the stack trace

Reply:
Done